docs/guide/serialization.rst

```diff
@@ -0,0 +1,310 @@
+
+On the other hand, a flag that changes code path in a way that might
+change the results of any operation would mean that the object cannot be a
+:class:`.GufeTokenizable`.
+
+.. _customjson:
+
+Supporting custom types in serialization
+----------------------------------------
+
+Occasionally, you may need to add support for a custom type that is not
+already known by GUFE. For example, there may be objects that are included
+in your GUFE object that come from another library, and are not GUFE
+tokenizables.
+
+When this happens, you will need to create and register a custom JSON
+codec. To do this, at a minimum you need to define a ``to_dict`` and
+``from_dict`` function for your type. You can then give the class of your
+type and those functions to a :class:`.JSONCodec`.  As an example, consider
+the code for our codec for ``pathlib.Path`` objects:
+
+.. code::
+
+    PATH_CODEC = JSONCodec(
+        cls=pathlib.Path,
+        to_dict=lambda p: {'path': str(p)},
+        from_dict=lambda dct: pathlib.Path(dct['path'])
+    )
+
+Here the ``to_dict`` and ``from_dict`` are lambdas. The ideas of these are
+the same as the ``GufeTokenizable._to_dict`` and
+``GufeTokenizable._from_dict`` described above.
+
+In this default setup, the codec recognizes your object type by doing an
+``isinstance`` check on the ``cls`` you gave it. It updates the dictionary
+that comes from your ``to_dict`` with the class and module of the object,
+as well as a key to mark this dictionary as coming from this codec. The
+key-value pairs that it adds make it so that the codec can recognize the
+dictionary when it is deserializes (decodes) the data.
+
+Details of how the object is recognized for encoding or how the dictionary
+is recognized for decoding can be changed by passing functions to the
+``is_my_obj`` or ``is_my_dict`` parameters of :class:`.JSONCodec`.
+
+Dumping arbitrary objects to JSON
+---------------------------------
+
+Any :class:`.GufeTokenizable` can be dumped to a JSON file using the custom
+JSON handlers. Given a :class:`.GufeTokenizable` called ``obj`` and a
+path-like called ``filename``, you can dump to JSON with this recipe:
+
+.. code::
+
+    import json
+    from gufe.tokenization import JSON_HANDLER
+    with open(filename, mode='w') as f:
+        json.dump(obj.to_dict(), f, cls=JSON_HANDLER.encoder)
+
+Similarly, you can reload the object with:
+
+.. code::
+
+    import json
+    from gufe.tokenization import JSON_HANDLER
+    with open(filename, mode='r') as f:
+        obj = json.load(f, cls=JSON_HANDLER.decoder)
+
+Note that these objects are not space-efficient: that is, if you have
+the same object in memory referenced by multiple objects (e.g., an identical
+``ProteinComponent`` in more than one ``ChemicalSystem``), then you will
+save multiple copies of its JSON representation.
+
+On reloading, tools that use the recommended ``from_dict`` method will undo
+do this duplication; see :ref:`gufe_memory_deduplication` for details.
```

Comment:
This reference is throwing a warning now

Reply:
fixed!